libraries/AP_LQR_Control/AP_LQR_Control.cpp

```diff
@@ -0,0 +1,409 @@
+        is_zero(lateral_accel_sea_level)) {
+        // Missing a sane input for calculating the limit, or the user has
+        // requested a straight scaling with altitude. This will always vary
+        // with the current altitude, but will at least protect the airframe
+        return radius * eas2tas_sq;
+    } else {
+        float sea_level_radius = sq(nominal_velocity_sea_level) / lateral_accel_sea_level;
+        if (sea_level_radius > radius) {
+            // If we've told the plane that its sea level radius is unachievable fallback to
+            // straight altitude scaling
+            return radius * eas2tas_sq;
+        } else {
+            // select the requested radius, or the required altitude scale, whichever is safer
+            return MAX(sea_level_radius * eas2tas_sq, radius);
+        }
+    }
+}
+
+bool AP_LQR_Control::reached_loiter_target(void)
+{
+    return _WPcircle;
+}
+
+/**
+   prevent indecision in our turning by using our previous turn
+   decision if we are in a narrow angle band pointing away from the
+   target and the turn angle has changed sign
+ */
+
+// update L1 control for waypoint navigation
+void AP_LQR_Control::update_waypoint(const struct Location &prev_WP, const struct Location &next_WP, float dist_min)
+{
+
+    struct Location _current_loc;
+    
+    uint32_t now = AP_HAL::micros();
+    float dt = (now - _last_update_waypoint_us) * 1.0e-6f;
+    if (dt > 0.1) {
+        dt = 0.1;
+    }
+    _last_update_waypoint_us = now;
+
+    // Get current position and velocity
+    if (_ahrs.get_position(_current_loc) == false) {
+        // if no GPS loc available, maintain last nav/target_bearing
+        _data_is_stale = true;
+        return;
+    }
+
+    Vector2f _groundspeed_vector = _ahrs.groundspeed_vector();
+
+    // update _target_bearing_cd
+    _target_bearing_cd = get_bearing_cd(_current_loc, next_WP);
+
+    //Calculate groundspeed
+    float groundSpeed = _groundspeed_vector.length();
+    if (groundSpeed < 0.1f) {
+        // use a small ground speed vector in the right direction,
+        // allowing us to use the compass heading at zero GPS velocity
+        groundSpeed = 0.1f;
+        _groundspeed_vector = Vector2f(cosf(get_yaw_rad()), sinf(get_yaw_rad())) * groundSpeed;
+    }
+
+    // Calculate the NE position of WP B relative to WP A
+    Vector2f AB = location_diff(prev_WP, next_WP);
+
+    // Check for AB zero length and track directly to the destination
+    // if too small
+    if (AB.length() < 1.0e-6f) {
+        AB = location_diff(_current_loc, next_WP);
+        if (AB.length() < 1.0e-6f) {
+            AB = Vector2f(cosf(get_yaw_rad()), sinf(get_yaw_rad()));
+        }
+    }
+    AB.normalize();
+
+    // Calculate the NE position of the aircraft relative to WP A
+    Vector2f A_air = location_diff(prev_WP, _current_loc);
+
+    // calculate distance to target track, for reporting
+    _crosstrack_error = A_air % AB;
```

Comment:
Why a cross product is treated as distance to target track?
![image](https://user-images.githubusercontent.com/7996972/50240065-802aa180-03cc-11e9-9206-7e2113682f6c.png)
I expect this to be copied from L1 code, but there it had somewhat different purpose and thus misleading name. Later in code it gets divided by another length to get sin of angle.
Strictly speaking cross product is not defined for 2D vectors, but for this case resulting meaning is area.

Reply:
Also the result can be treated as a measure of how parallel or perpendicular two vectors are, but definitely not term "d" as for this method.

Reply:
Can I divide the area obtained by the length of AB for calculating d?

Reply:
Seems to be so